vyper/ast/folding.py

```diff
@@ -0,0 +1,119 @@
+from typing import (
+    Union,
+)
+
+from vyper.ast import (
+    nodes as vy_ast,
+)
+from vyper.exceptions import (
+    InvalidType,
+)
+
+
+def fold(vyper_ast_node: vy_ast.Module) -> None:
+    """
+    Perform literal folding operations on a Vyper AST.
+
+    Arguments
+    ---------
+    vyper_ast_node : Module
+        Top-level Vyper AST node.
+    """
+    changed_nodes = 1
+    while changed_nodes:
+        changed_nodes = 0
+        changed_nodes += replace_literal_ops(vyper_ast_node)
+        changed_nodes += replace_subscripts(vyper_ast_node)
+
+
+def replace_literal_ops(vyper_ast_node: vy_ast.Module) -> int:
+    """
+    Find and evaluate operation and comparison nodes within the Vyper AST,
+    replacing them with Constant nodes where possible.
+
+    Arguments
+    ---------
+    vyper_ast_node : Module
+        Top-level Vyper AST node.
+    """
+    changed_nodes = 0
+
+    node_types = (vy_ast.BoolOp, vy_ast.BinOp, vy_ast.UnaryOp, vy_ast.Compare)
+    for node in vyper_ast_node.get_descendants(node_types, reverse=True):
+        try:
+            new_node = node.evaluate()
+        except InvalidType:
+            continue
+
+        changed_nodes += 1
+        vyper_ast_node.replace_in_tree(node, new_node)
+
+    return changed_nodes
+
+
+def replace_subscripts(vyper_ast_node: vy_ast.Module) -> int:
+    """
+    Find and evaluate Subscript nodes within the Vyper AST, replacing them with
+    Constant nodes where possible.
+
+    Arguments
+    ---------
+    vyper_ast_node : Module
+        Top-level Vyper AST node.
+    """
+    changed_nodes = 0
+
+    for node in vyper_ast_node.get_descendants(vy_ast.Subscript, reverse=True):
+        try:
+            new_node = node.evaluate()
+        except InvalidType:
+            continue
+
+        changed_nodes += 1
+        vyper_ast_node.replace_in_tree(node, new_node)
+
+    return changed_nodes
+
+
+def _replace(old_node, new_node):
+    if isinstance(new_node, vy_ast.Constant):
+        return new_node.from_node(old_node, value=new_node.value)
+    elif isinstance(new_node, vy_ast.List):
+        list_values = [_replace(old_node, i) for i in new_node.elts]
+        return new_node.from_node(old_node, elts=list_values)
```

Comment:
Seriously, what the hell is an `elt`? Can we pick a different name?

Reply:
Short for elements. It's a field name we're inheriting from python that holds all the values within a list or tuple.

We could rename it for sure, but probably belongs in a different PR.